### DISCUSSION TITLE
Shortest Path between Nodes in a Graph

### DISCUSSION THREAD

**Discussion by malikhizar** [networkx/networkx#5874] [Q&A]
I am are using the NetworkX library for calculating the length between any two nodes in a given graph data. Using the 'edges' file, where one column is the source node and the other column is the target node. I am using 'nx.all_pairs_shortest_path_length(G)'  function to path computation for graph G. The problem is that when it finds a path (of any length) between any two nodes in G, it returns that path, however, when this function cannot find a path between two given nodes (it might be some isolated or unconnected nodes in G), it gives an error, 'cannot find the key XYZ'. How to handle this error or which function to use from the library?

**Comment by rossbar** [comment]
I'm not sure I'm understanding the question correctly, but maybe a general example will help.

`dict(nx.all_pairs_shortest_path(G))` will create a dict-of-dict structure where the outer key represents the "source" node, the inner key the "target" node, and the value the shortest path between them. Thus the easiest way to look up the shortest path between two nodes in a graph `G` is to do lookups in this dict. For example:

```python
>>> G = nx.complete_graph(5)  # shortest path between two different nodes is the edge between them
>>> G.add_node(6)  # a disconnected node
>>> shortest_paths = dict(nx.all_pairs_shortest_paths(G))
>>> shortest_paths
{0: {0: [0], 1: [0, 1], 2: [0, 2], 3: [0, 3], 4: [0, 4]},
 1: {1: [1], 0: [1, 0], 2: [1, 2], 3: [1, 3], 4: [1, 4]},
 2: {2: [2], 0: [2, 0], 1: [2, 1], 3: [2, 3], 4: [2, 4]},
 3: {3: [3], 0: [3, 0], 1: [3, 1], 2: [3, 2], 4: [3, 4]},
 4: {4: [4], 0: [4, 0], 1: [4, 1], 2: [4, 2], 3: [4, 3]},
 6: {6: [6]}}
```

Now, to get the shortest path between any two nodes, we do a simple lookup in the shortest paths dict:

```python
>>> shortest_paths[2][3]
[2, 3]
```

Easy!

IIUC correctly, your question is about what happens when there *isn't* a path between any two nodes. That depends on how you do the lookup. With the standard (i.e. `__getitem__`) lookup, you'll get a `KeyError` if there is no path between two nodes.

```python
>>> shortest_paths[0][6]  # Remember: node 6 is not connected
Traceback (most recent call last)
   ...
KeyError: 6
```

How you handle this information is up to you. For example, if you want to ignore failed lookups, you could put this in a try-except statement:

```python
try:
    shortest_paths[u][v]
except KeyError:
    # Whatever processing you want to do when a path doesn't exist or a node is not in G.
```

Alternatively, if you want to e.g. return a sentinel value when a path doesn't exist, it might be easier to use the `dict.get` method:

```python
>>> path = shortest.paths.get(0).get(6)  # .get returns `None` (by default) when a key is not found
>>> path is None
True
```

As you can see, there are many different ways you can handle the case when there is no path between two nodes - which is best depends on your specific application. Hopefully this addresses your question!
